examples/advanced/ocus_explanations.py

```diff
@@ -0,0 +1,302 @@
+            "derived": list(derived),
+            "cost": sum(cost(con) for con in ocus_expl)
+        }
+
+        explanation_sequence.append(explanation)
+
+    # return the list of original (non-flattened) constraints
+    return explanation_sequence
+
+def explain_one_step_ocus(hard, cost, Iend, I, solver="ortools", verbose=False):
+    """
+        Optimal Constrained Unsatisfiable Subsets (OCUS) for CSP explanations [1]
+
+        explain_one_step_ocus relies on the implicit hitting set duality between
+        MUSes and MCSes for a given formula F [2, 3]:
+
+            - A set S \subseteq of F is an MCS of F iff it is a minimal hitting set
+             of MUSs(F).
+            - A set S \subseteq F is a MUS of F iff it is a minimal hitting set
+             of MCSs(F).
+
+        Builds a MIP model for computing minimal (optimal) hitting sets. Repeatedly
+        checks for satisfiability until UNSAT, which means an OCUS is found.
+        If SAT, and sat solver computes a model. The complement w.r.t F is
+        computed and added as a new set-to-hit.
+
+        MIP MODEL
+        ---------
+
+        The constrained optimal hitting set is described by:
+
+            - x_l={0,1} is a boolean decision variable if the literal is inside the
+                        hitting set or not.
+            - w_l=f(l) is the cost assigned to having the literal in the hitting
+                        set.
+            - c_lj={0,1} is 1 (0) if the literal l is (not) present in the set-to-hit j.
+
+        Objective:
+                min sum(x_l * w_l) over all l in I + (-Iend \ -I)
+
+        Subject to:
+            (1) sum x_l * c_lj >= 1 for all hitting sets j.
+                = The hitting set must hit all sets-to-hit.
+
+            (2) sum x_l == 1 for l in (-Iend \ -I)
+                = exactly 1 literal explained at a time
+
+        Args
+        ----
+
+            hard (list[cpmpy.Expression]): Hard constraints
+
+            Iend (set): Cautious consequence, the set of literals that hold in
+                        all models.
+
+            I (set): partial interpretation (subset of Iend).
+
+        [1] Gamba, E., Bogaerts, B., & Guns, T. (8 2021). Efficiently Explaining CSPs
+        with Unsatisfiable Subset Optimization. In Z.-H. Zhou (Red), Proceedings of the
+        Thirtieth International Joint Conference on Artificial Intelligence,
+        IJCAI-21 (bll 1381–1388). doi:10.24963/ijcai.2021/191.
+
+        [2] Liffiton, M. H., & Sakallah, K. A. (2008). Algorithms for computing minimal
+        unsatisfiable subsets of constraints. Journal of Automated Reasoning, 40(1), 1-33.
+
+        [3] Reiter, R. (1987). A theory of diagnosis from first principles.
+        Artificial intelligence, 32(1), 57-95.
+    """
+
+    sat = SolverLookup.lookup(solver)(Model(hard))
+
+    ## OPT Model: Variables
+    hs_vars = boolvar(shape=len(Iend))
```

Comment:
somehow, it would be great to make it clearer that 'sat' here is over a set of variables (Iend I guess), and that hs_model will be over another; and then you maintain the mapping...

but a clearer separation would be nice.

perhaps Iend -> sat_vars; sat -> sat_model
then hs_vars and hs_model as now. And then, your maps can be called varmap_sat_hs and varmap_hs_sat or so?